File: src/StreamVue.Player/Services/PremiumAccessPolicy.cs
namespace StreamVue.Player.Services;

public enum StreamVueDistributionMode
{
    Personal,
    Store,
    Unknown
}

public enum PremiumAccessState
{
    Included,
    Verified,
    Unavailable
}

public sealed record PremiumAccessSnapshot(
    string ContractVersion,
    string FeatureId,
    StreamVueDistributionMode DistributionMode,
    PremiumAccessState AccessState,
    string Acquisition,
    string ReceiptVerification,
    string? ProductId = null)
{
    public bool CanUseMediaCenters => AccessState is PremiumAccessState.Included or PremiumAccessState.Verified;

    public string BadgeText => AccessState switch
    {
        PremiumAccessState.Included => "PERSONAL BUILD • INCLUDED",
        PremiumAccessState.Verified => "PREMIUM • VERIFIED",
        _ => "PREMIUM • STORE LOCKED"
    };

    public string Explanation => AccessState switch
    {
        PremiumAccessState.Included => "Plex and Emby are included in this personal build.",
        PremiumAccessState.Verified => "A one-time store purchase was verified for this device account.",
        _ => "A verified one-time store purchase is required. Store purchase verification is not connected in this build."
    };
}

public static class PremiumAccessPolicy
{
    public const string ContractVersion = "1.0";
    public const string MediaCentersFeatureId = "personal-media-centers";

    public static PremiumAccessSnapshot Current
    {
        get
        {
#if STREAMVUE_STORE_BUILD
            return Evaluate("store", hasVerifiedStorePurchase: false);
#else
            return Evaluate("personal", hasVerifiedStorePurchase: false);
#endif
        }
    }

    public static PremiumAccessSnapshot Evaluate(
        string? distributionMode,
        bool hasVerifiedStorePurchase,
        string? productId = null)
    {
        var mode = ParseDistributionMode(distributionMode);
        if (mode == StreamVueDistributionMode.Personal)
        {
            return new PremiumAccessSnapshot(
                ContractVersion,
                MediaCentersFeatureId,
                mode,
                PremiumAccessState.Included,
                "included",
                "not-required");
        }

        var normalizedProductId = NormalizeProductId(productId);
        if (mode == StreamVueDistributionMode.Store &&
            hasVerifiedStorePurchase &&
            normalizedProductId is not null)
        {
            return new PremiumAccessSnapshot(
                ContractVersion,
                MediaCentersFeatureId,
                mode,
                PremiumAccessState.Verified,
                "one-time",
                "verified",
                normalizedProductId);
        }

        return new PremiumAccessSnapshot(
            ContractVersion,
            MediaCentersFeatureId,
            mode,
            PremiumAccessState.Unavailable,
            "one-time",
            "unavailable");
    }

    public static void RequireMediaCenters(PremiumAccessSnapshot access)
    {
        if (!access.CanUseMediaCenters) throw new InvalidOperationException(access.Explanation);
    }

    private static StreamVueDistributionMode ParseDistributionMode(string? value) =>
        value?.Trim().ToLowerInvariant() switch
        {
            "personal" => StreamVueDistributionMode.Personal,
            "store" => StreamVueDistributionMode.Store,
            _ => StreamVueDistributionMode.Unknown
        };

    private static string? NormalizeProductId(string? value)
    {
        var candidate = value?.Trim();
        if (string.IsNullOrWhiteSpace(candidate) || candidate.Length is < 3 or > 256) return null;
        return candidate.All(character => char.IsAsciiLetterOrDigit(character) || character is '.' or '_' or '-')
            ? candidate
            : null;
    }
}

File: platforms/android/app/src/main/java/com/streamvue/player/premium/PremiumAccessPolicy.kt
package com.streamvue.player.premium

import com.streamvue.player.BuildConfig

enum class StreamVueDistributionMode {
    Personal,
    Store,
    Unknown
}

enum class PremiumAccessState {
    Included,
    Verified,
    Unavailable
}

data class PremiumAccessSnapshot(
    val contractVersion: String,
    val featureId: String,
    val distributionMode: StreamVueDistributionMode,
    val accessState: PremiumAccessState,
    val acquisition: String,
    val receiptVerification: String,
    val productId: String? = null
) {
    val canUseMediaCenters: Boolean
        get() = accessState == PremiumAccessState.Included || accessState == PremiumAccessState.Verified

    val badgeText: String
        get() = when (accessState) {
            PremiumAccessState.Included -> "PERSONAL BUILD • INCLUDED"
            PremiumAccessState.Verified -> "PREMIUM • VERIFIED"
            PremiumAccessState.Unavailable -> "PREMIUM • STORE LOCKED"
        }

    val explanation: String
        get() = when (accessState) {
            PremiumAccessState.Included -> "Plex and Emby are included in this personal build."
            PremiumAccessState.Verified -> "A one-time store purchase was verified for this device account."
            PremiumAccessState.Unavailable ->
                "A verified one-time store purchase is required. Store purchase verification is not connected in this build."
        }

    fun requireMediaCenters() {
        check(canUseMediaCenters) { explanation }
    }
}

object PremiumAccessPolicy {
    const val ContractVersion = "1.0"
    const val MediaCentersFeatureId = "personal-media-centers"

    fun current(): PremiumAccessSnapshot = evaluate(
        distributionMode = BuildConfig.DISTRIBUTION_MODE,
        hasVerifiedStorePurchase = false
    )

    fun evaluate(
        distributionMode: String?,
        hasVerifiedStorePurchase: Boolean,
        productId: String? = null
    ): PremiumAccessSnapshot {
        val mode = when (distributionMode?.trim()?.lowercase()) {
            "personal" -> StreamVueDistributionMode.Personal
            "store" -> StreamVueDistributionMode.Store
            else -> StreamVueDistributionMode.Unknown
        }
        if (mode == StreamVueDistributionMode.Personal) {
            return PremiumAccessSnapshot(
                contractVersion = ContractVersion,
                featureId = MediaCentersFeatureId,
                distributionMode = mode,
                accessState = PremiumAccessState.Included,
                acquisition = "included",
                receiptVerification = "not-required"
            )
        }

        val normalizedProductId = productId?.trim()?.takeIf(::isValidProductId)
        if (mode == StreamVueDistributionMode.Store &&
            hasVerifiedStorePurchase &&
            normalizedProductId != null
        ) {
            return PremiumAccessSnapshot(
                contractVersion = ContractVersion,
                featureId = MediaCentersFeatureId,
                distributionMode = mode,
                accessState = PremiumAccessState.Verified,
                acquisition = "one-time",
                receiptVerification = "verified",
                productId = normalizedProductId
            )
        }

        return PremiumAccessSnapshot(
            contractVersion = ContractVersion,
            featureId = MediaCentersFeatureId,
            distributionMode = mode,
            accessState = PremiumAccessState.Unavailable,
            acquisition = "one-time",
            receiptVerification = "unavailable"
        )
    }

    private fun isValidProductId(value: String): Boolean =
        value.length in 3..256 && value.all {
            it in 'a'..'z' || it in 'A'..'Z' || it in '0'..'9' || it == '.' || it == '_' || it == '-'
        }
}
